modules/volume/volume.go

```diff
@@ -247,37 +307,243 @@ func (m *Module) worker() {
+	if err != nil {
+		return nil, err
+	}
+
+	locator := bus.Object("org.PulseAudio1", "/org/pulseaudio/server_lookup1")
+	path, err := locator.GetProperty("org.PulseAudio.ServerLookup1.Address")
+	if err != nil {
+		return nil, err
+	}
+
+	return dialAndAuth(path.Value().(string))
+}
+
+func createPulseModule(sinkName string) *Module {
+	m := &paModule{
+		sinkName: sinkName,
+	}
+	m.impl = m
+	m.init()
+	return &m.Module
+}
+
+// Create a PulseAudio volume module that follows the default sink.
+func DefaultSink() *Module {
+	return createPulseModule("")
+}
+
+// Create a PulseAudio volume module for a named sink.
+func Sink(sinkName string) *Module {
+	return createPulseModule(sinkName)
+}
+
+func (m *paModule) SetVolume(newVol int64) {
+	vol, _ := m.currentVolume.Get()
+	if vol == nil {
+		return
+	}
+	v := vol.(Volume)
+	if newVol > v.Max {
+		newVol = v.Max
+	}
+	if newVol < v.Min {
+		newVol = v.Min
+	}
+	call := m.sink.Call("org.freedesktop.DBus.Properties.Set", 0,
+		"org.PulseAudio.Core1.Device", "Volume", dbus.MakeVariant([]uint32{uint32(newVol)}))
+
+	if call.Err != nil {
+		fmt.Fprintf(os.Stderr, "Could not set volume: %v\n", call.Err)
+	}
+
+	m.refresh()
+}
+
+func (m *paModule) SetMuted(muted bool) {
+	call := m.sink.Call("org.freedesktop.DBus.Properties.Set", 0,
+		"org.PulseAudio.Core1.Device", "Mute", dbus.MakeVariant(muted))
+	if call.Err != nil {
+		fmt.Fprintf(os.Stderr, "Could not set mute: %v\n", call.Err)
+	}
+	m.refresh()
+}
+
+func (m *paModule) listen(signal string, objects ...dbus.ObjectPath) error {
+	call := m.core.Call("org.PulseAudio.Core1.ListenForSignal", 0, "org.PulseAudio.Core1."+signal, objects)
+	return call.Err
+}
+
+func (m *paModule) openSink(sink dbus.ObjectPath) error {
+	m.sink = m.conn.Object("org.PulseAudio.Core1.Sink", sink)
+
+	if err := m.listen("Device.VolumeUpdated", sink); err != nil {
+		return err
+	}
+	if err := m.listen("Device.MuteUpdated", sink); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *paModule) openSinkByName(name string) error {
+	var path dbus.ObjectPath
+
+	err := m.core.Call("org.PulseAudio.Core1.GetSinkByName", 0, name).Store(&path)
+	if err != nil {
+		return err
+	}
+
+	return m.openSink(path)
+}
+
+func (m *paModule) openFallbackSink() error {
+	path, err := m.core.GetProperty("org.PulseAudio.Core1.FallbackSink")
+	if err != nil {
+		return err
+	}
+
+	return m.openSink(path.Value().(dbus.ObjectPath))
+}
+
+func (m *paModule) refresh() error {
+	v := Volume{}
+	v.Min = 0
+
+	max, err := m.sink.GetProperty("org.PulseAudio.Core1.Device.BaseVolume")
+	if err != nil {
+		return err
+	}
+	v.Max = int64(max.Value().(uint32))
+
+	vol, err := m.sink.GetProperty("org.PulseAudio.Core1.Device.Volume")
+	if err != nil {
+		return err
+	}
+
+	// Take the volume as the average across all channels.
+	var totalVol int64
+	channels := vol.Value().([]uint32)
+	for _, ch := range channels {
+		totalVol += int64(ch)
+	}
+	v.Vol = totalVol / int64(len(channels))
```

Comment:
Do we need to store the number of channels so we can send the right number of uint32s when setting the volume, or does sending one implicitly set all channels?

Reply:
For Pulse, setting the volume to a length-1 array is documented to change the volume for all channels:

> [There are two ways to adjust the volume. You can either adjust the overall volume by giving a single-value array, or you can precisely control the individual channels by passing an array containing a value for each channel.](https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/DBus/Device/#volume)

Reply:
:+1: